### PR TITLE
FieldId and Listindex should not be exclusive

### PR DESCRIPTION
**Pull request by yunhanw-google**
#### Problem
We would only support toplevel attributes and list items inside toplevel attribute
#7055

#### Change overview
--Replace ListIndex with kFieldIdListIndexValid
--Support two situation in attribute path, one is FieldId only, another
is FieldId + ListIndex.
--Update ClusterInfo to support kFieldIdListIndexValid

#### Testing
Add unit test to cover FieldId only and FieldId+ListIndex

### CODE REVIEW

**Review by msandstedt — APPROVED** [review]

**Comment by woody-apple** [comment]
@bzbarsky-apple ?

File: src/app/ReadClient.cpp [hunk 1]
```diff
@@ -187,6 +163,38 @@ CHIP_ERROR ReadClient::SendReadRequest(NodeId aNodeId, Transport::AdminId aAdmin
...
+    {
+        AttributePath::Builder attributePathBuilder = attributePathListBuilder.CreateAttributePathBuilder();
+        attributePathBuilder.NodeId(apAttributePathParamsList[index].mNodeId)
+            .EndpointId(apAttributePathParamsList[index].mEndpointId)
+            .ClusterId(apAttributePathParamsList[index].mClusterId);
+        if (apAttributePathParamsList[index].mFlags == AttributePathFlags::kFieldIdValid)
+        {
+            attributePathBuilder.FieldId(apAttributePathParamsList[index].mFieldId);
+        }
+        else if (apAttributePathParamsList[index].mFlags == AttributePathFlags::kFieldIdListIndexValid)
```
Comment (Damian-Nordic) [inline src/app/ReadClient.cpp:181]: It's not clear to me why mFlags is `BitFlags` while you seem to use it like an enum. Is it possible to have only the list index without the field id? If yes check both bits independently without the `else` keword. If not, you don't need `BitFlags` and just switch to a bare enum class.

Comment (yunhanw-google) [reply src/app/ReadClient.cpp]: agree, just use BitFlags to bit comparison, thanks

**Review by andy31415 — APPROVED** [review]

File: src/app/tests/TestReadInteraction.cpp [hunk 2]
```diff
@@ -72,6 +73,48 @@ void TestReadInteraction::GenerateReportData(nlTestSuite * apSuite, void * apCon
     err = reportDataBuilder.Init(&writer);
     NL_TEST_ASSERT(apSuite, err == CHIP_NO_ERROR);
 
+    AttributeDataList::Builder attributeDataListBuilder = reportDataBuilder.CreateAttributeDataListBuilder();
```
Comment (bzbarsky-apple) [inline src/app/tests/TestReadInteraction.cpp:76]: I don't understand why we we bothering making changes to this function; is its output ever used for anything?

**Review by bzbarsky-apple — APPROVED** [review]

**Review by woody-apple — CHANGES_REQUESTED** [review]
Requesting changes for @bzbarsky-apple

File: src/app/ReadClient.cpp [hunk 3]
```diff
@@ -187,6 +163,34 @@ CHIP_ERROR ReadClient::SendReadRequest(NodeId aNodeId, Transport::AdminId aAdmin
...
+        AttributePath::Builder attributePathBuilder = attributePathListBuilder.CreateAttributePathBuilder();
+        attributePathBuilder.NodeId(apAttributePathParamsList[index].mNodeId)
+            .EndpointId(apAttributePathParamsList[index].mEndpointId)
+            .ClusterId(apAttributePathParamsList[index].mClusterId);
+        if (apAttributePathParamsList[index].mFlags.Has(AttributePathParams::Flags::kFieldIdValid))
+        {
+            attributePathBuilder.FieldId(apAttributePathParamsList[index].mFieldId);
+        }
+
+        if (apAttributePathParamsList[index].mFlags.Has(AttributePathParams::Flags::kListIndexValid))
```
[the remark this answers is not in the document]
Comment (yunhanw-google) [reply src/app/ReadClient.cpp:182]: thanks for catching this issue, Just fix it.

File: src/app/tests/TestReadInteraction.cpp (hunk 2 above)
Comment (yunhanw-google) [reply src/app/tests/TestReadInteraction.cpp:76]: yep, the report data would be processed inside ProcessReportData, which is tested , then report.CheckSchemaValidity and  ProcessAttributeDataList is exercised  inside ProcessReportData

**Comment by woody-apple** [comment]
@bzbarsky-apple , are your comments resolved?

**Review by woody-apple — APPROVED** [review]

**Pull request merged**